### PR TITLE
Quote reserved identifiers

### LINKED ISSUE #13483: SHOW CREATE TABLE and SHOW COLUMNS return invalid type when column/field name is a reserved keyword
If name of a columns or a field is a reserved keyword, Trino should quote it when printing type in `SHOW COLUMNS FROM table`, but it does not. It also does not do it for column names in `SHOW CREATE TABLE`.

E.g.
```
CREATE TABLE test_type (
    x1 ROW("exists" BOOLEAN),
    "exists" VARCHAR
 )
```

`> SHOW CREATE TABLE test_type;`
```
CREATE TABLE test_type (
    x1 ROW("exists" boolean),
    exists varchar
 )
```
Note missing quotes on the column name. Running this create table statement will result in an error `line 3:5: mismatched input 'exists'. Expecting: 'LIKE', <identifier>`

`> SHOW COLUMNS FROM test_type;`
```
 Column |        Type         | Extra | Comment
--------+---------------------+-------+---------
 x1     | row(exists boolean) |       |
 exists | varchar             |       |
(2 rows)
```
Note missing quotes on the column name and field name.

### PR DESCRIPTION
**Pull request by nevillelyh**
## Description

Fix #13483, quote reserved keywords in `SHOW CREATE TABLE`

> Is this change a fix, improvement, new feature, refactoring, or other?

Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core query engine

> How would you describe this change to a non-technical end user or system administrator?

## Related issues, pull requests, and links

Fixes #13483

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by nevillelyh** [comment]
I got this working with the `MemoryQueryRunner`. However there doesn't seem to be any tests for the SQL formatters, so it's hard to gauge the impact. Any idea how to proceed?

```
trino> CREATE TABLE memory.default.test_type (
    ->     x1 ROW("exists" BOOLEAN),
    ->     "exists" VARCHAR
    -> );
CREATE TABLE
trino> show create table memory.default.test_type;
              Create Table
-----------------------------------------
 CREATE TABLE memory.default.test_type (
    x1 ROW("exists" boolean),
    "exists" varchar
 )
(1 row)
```

File: core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java [hunk 1]
```diff
@@ -363,7 +363,7 @@ protected String visitExists(ExistsPredicate node, Void context)
         @Override
         protected String visitIdentifier(Identifier node, Void context)
         {
-            if (!node.isDelimited()) {
+            if (!node.isDelimited() && !ReservedIdentifiers.reservedIdentifiers().contains(node.getValue().toUpperCase())) {
```
Comment (findepi) [inline core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:366]: `reservedIdentifiers` is expensive to call

Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:366]: Yeah I realized. Just want to get an opinion if this is the right approach first.
To optimize, should we make `ReservedIdentifiers.sqlKeywords` & `reservedIdentifiers` return static instance instead?

Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java]: Added static members for these 2 methods.

File: core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java [hunk 2]
```diff
@@ -363,7 +363,7 @@ protected String visitExists(ExistsPredicate node, Void context)
         @Override
         protected String visitIdentifier(Identifier node, Void context)
         {
-            if (!node.isDelimited()) {
+            if (!node.isDelimited() && !ReservedIdentifiers.reservedIdentifiers().contains(node.getValue().toUpperCase(Locale.ENGLISH))) {
```
Comment (findepi) [inline core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:366]: import `ReservedIdentifiers.reservedIdentifiers()` statically

**Review by findepi — APPROVED** [review]
Can you please squash commits?

**Comment by findepi** [comment]
@nevillelyh would you be able to add some `SHOW CREATE TABLE` test?

**Comment by nevillelyh** [comment]
> @nevillelyh would you be able to add some `SHOW CREATE TABLE` test?

I briefly looked at tests.

File: core/trino-parser/src/test/java/io/trino/sql/parser/TreeNodes.java [hunk 3]
```diff
@@ -35,7 +35,7 @@
 
 import static java.util.Collections.emptyList;
 
-class TreeNodes
+public class TreeNodes
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/parser/TreeNodes.java:38]: This class is meant to be private for testing the SQL parser. Don't make it public.

File: core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java [hunk 4]
```diff
@@ -44,6 +44,27 @@
...
+            if (matcher.matches()) {
+                names.add(matcher.group(1));
+            }
+        }
+        SQL_KEYWORDS = names.build();
+        RESERVED_IDENTIFIERS = SQL_KEYWORDS.stream()
+                .filter(ReservedIdentifiers::reserved)
+                .sorted()
+                .collect(toImmutableSet());
+    }
```
Comment (martint) [inline core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:66]: This change should go in a separate commit. It's tangentially related to the main fix in this PR (needed for performance), but it's not an essential part of the change.

**Review by martint — CHANGES_REQUESTED** [review]

File: core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java (hunk 4 above)
Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:66]: Moved this to #13739

File: core/trino-parser/src/test/java/io/trino/sql/parser/TreeNodes.java (hunk 3 above)
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/parser/TreeNodes.java]: The `*Formatter` classes are in `io.trino.sql`, one level up from `parser` so I created the test in the same package.

I removed the dependency & re-impl:ed `simpleType` & `location` inside the test for now. We might end up with more duplications though if we add more formatter tests like this.

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private static final String TABLE_NAME = "table_name";
+    private static final String COLUMN_NAME = "column_name";
+    private static final String RESERVED_TABLE_NAME = "exists";
+    private static final String RESERVED_COLUMN_NAME = "exists";
+    private static final String CREATE_TABLE = "CREATE TABLE %s (\n   %s VARCHAR\n)";
+
+    @Test
+    public void testCreateTable()
+    {
+        assertEquals(String.format(CREATE_TABLE, TABLE_NAME, COLUMN_NAME), createTable(TABLE_NAME, COLUMN_NAME));
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:41]: `createTable` is misleading. It's not obvious that it's purpose is to do the bulk of the work -- i.e., formatting a CreateTable node with a specific shape (see last paragraph in this comment).

I would restructure the test as follows (also, using AssertJ assertions):

```java
assertThat(formatSql(createTable(tableName, columnName)))
    .isEqualTo("CREATE TABLE %s (   %s VARCHAR)".formatted(tableName, columnName));
```

and 

```java
private CreateTable createTable(String tableName, String columnName)
{
     ...
     return new CreateTable(...);
}
```

However, I might even just inline the whole createTable method in the caller. There's a tight dependency between what the createTable method does (based on the arbitrary choice of having a single VARCHAR column) and the expected string that comes out of formatting the returned CreateTable node. However, those concerns are scattered throughout the class in the current implementation (static fields, test method, createTable method)

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import static io.trino.sql.QueryUtil.identifier;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TestSqlFormatter
+{
+    private static final String TABLE_NAME = "table_name";
+    private static final String COLUMN_NAME = "column_name";
+    private static final String RESERVED_TABLE_NAME = "exists";
+    private static final String RESERVED_COLUMN_NAME = "exists";
+    private static final String CREATE_TABLE = "CREATE TABLE %s (\n   %s VARCHAR\n)";
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: Inlined them after moving everything into the method.

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java (hunk 5 above)
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: - Moved everything into a single method
- Switched to AssertJ
- Kept `String.format` as `trino-parser` is still targeting JDK 8
- Made `createTable` a lambda (Am I overdoing it?)

File: core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java (hunk 4 above)
Comment (findepi) [reply core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java]: @nevillelyh let's have this as separate commit, but please keep this within this PR.

Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:66]: @findepi done

Comment (findepi) [reply core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:66]: Can you please also change the order of commits? `Make reservedIdentifiers & sqlKeywords static` should be first.
Also, add some explanation there, like "This makes reservedIdentifiers() and sqlKeywords() constant time operations."

Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:66]: Fixed!

Comment (findepi) [reply core/trino-parser/src/main/java/io/trino/sql/ReservedIdentifiers.java:66]: thanks

**Comment by kasiafi** [comment]
@martint, BTW, regarding the method `requiresDelimiting(String identifier)` in `TypeSignatureTranslator` I mentioned in this comment: https://github.com/trinodb/trino/pull/13611#discussion_r953616631, I believe that the "quoting" logic should be moved to the `ExpressionFormatter`, and removed from the `TypeSignatureTranslator`. We should avoid quoting identifiers when not necessary (like in the formatter, to produce a valid SQL), since it changes their semantics.

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+public class TestSqlFormatter
+{
+    @Test
+    public void testCreateTable()
+    {
+        BiFunction<String, String, CreateTable> createTable = (tableName, columnName) -> {
+            NodeLocation location = new NodeLocation(1, 1);
+            Identifier identifier = new Identifier(location, "VARCHAR", false);
+            return new CreateTable(
+                    QualifiedName.of(tableName),
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:41]: Done!

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                    ImmutableList.of(new ColumnDefinition(
+                            identifier(columnName),
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:43]: Done!

File: core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java [hunk 9]
```diff
@@ -363,7 +364,7 @@ protected String visitExists(ExistsPredicate node, Void context)
         @Override
         protected String visitIdentifier(Identifier node, Void context)
         {
-            if (!node.isDelimited()) {
+            if (!node.isDelimited() && !reservedIdentifiers().contains(node.getValue().toUpperCase(Locale.ENGLISH))) {
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:367]: Done!

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.trino.sql.QueryUtil.identifier;
+import static io.trino.sql.SqlFormatter.formatSql;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestSqlFormatter
+{
+    @Test
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:34]: Done!

File: core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java:367]: Fixed. I must have missed something and didn't get `reserved` to work on my first attempt.

[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java]: Yeah, they could be now extracted as a separate PR.

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java]: > However the optimizations for `reservedIdentifiers()` & `sqlKeywords()` are no longer relevant to this PR after the change. Do we still want to them? Revive #13739 maybe?

Then they are no longer useful, since other usages of these methods do not care about their cost.

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        // Non-reserved keyword
+        assertThat(formatSql(new Identifier("analyze"))).isEqualTo("analyze");
+        assertThat(formatSql(new Identifier("\"analyze\""))).isEqualTo("\"\"\"analyze\"\"\"");
+
+        // ANSI-compliant string
+        assertThat(formatSql(new Identifier("account"))).isEqualTo("account");
+        assertThat(formatSql(new Identifier("\"account\""))).isEqualTo("\"\"\"account\"\"\"");
+
+        // Non-ANSI quoted string
+        assertThat(formatSql(new Identifier("\"1\""))).isEqualTo("\"\"\"1\"\"\"");
```
Comment (kasiafi) [inline core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:49]: I meant testing the same word as quoted and unquoted identifier, like below.
```suggestion
        // Reserved keyword
        assertThat(formatSql(new Identifier("exists", false))).isEqualTo("\"exists\"");
        assertThat(formatSql(new Identifier("exists", true))).isEqualTo("\"exists\"");

        // Non-reserved keyword
        assertThat(formatSql(new Identifier("analyze", false))).isEqualTo("analyze");
        assertThat(formatSql(new Identifier("analyze", true))).isEqualTo("\"analyze\"");

        // ANSI-compliant string
        assertThat(formatSql(new Identifier("account", false))).isEqualTo("account");
        assertThat(formatSql(new Identifier("account", true))).isEqualTo("\"account\"");

        // Non-ANSI quoted string
        assertThat(formatSql(new Identifier("1", true))).isEqualTo("\"1\"");
```

However it is a good idea to test quotes as part of the identifier name. If you want to do this, please add another test method.

**Review by kasiafi — COMMENTED** [review]
LGTM % comments

File: core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java [hunk 12]
```diff
@@ -363,11 +364,11 @@ protected String visitExists(ExistsPredicate node, Void context)
...
         protected String visitIdentifier(Identifier node, Void context)
         {
-            if (!node.isDelimited()) {
-                return node.getValue();
+            if (node.isDelimited() || reserved(node.getValue())) {
+                return '"' + node.getValue().replace("\"", "\"\"") + '"';
             }
             else {
-                return '"' + node.getValue().replace("\"", "\"\"") + '"';
+                return node.getValue();
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java]: Done

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @Test
+    public void testCreateTable()
+    {
+        BiFunction<String, String, CreateTable> createTable = (tableName, columnName) -> {
+            NodeLocation location = new NodeLocation(1, 1);
+            Identifier identifier = new Identifier(location, "VARCHAR", false);
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: Done

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+            return new CreateTable(
+                    QualifiedName.of(ImmutableList.of(new Identifier(tableName))),
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: Done

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java (hunk 11 above)
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: Done & added another method for quotes as part of identifier name.

File: core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nevillelyh) [reply core/trino-parser/src/main/java/io/trino/sql/ExpressionFormatter.java]: Removed the optimization commit.

**Review by kasiafi — APPROVED** [review]

**Comment by kasiafi** [comment]
@martint could you please take a look?

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        // Reserved keyword
+        assertThat(formatSql(new Identifier("exists", false))).isEqualTo("\"exists\"");
+        assertThat(formatSql(new Identifier("exists", true))).isEqualTo("\"exists\"");
+
+        // Non-reserved keyword
+        assertThat(formatSql(new Identifier("analyze", false))).isEqualTo("analyze");
+        assertThat(formatSql(new Identifier("analyze", true))).isEqualTo("\"analyze\"");
+
+        // ANSI-compliant string
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:44]: This is not a SQL string -- it's an identifier. Strings are delimited using single quotes.

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+        assertThat(formatSql(new Identifier("account", false))).isEqualTo("account");
+        assertThat(formatSql(new Identifier("account", true))).isEqualTo("\"account\"");
+
+        // Non-ANSI quoted string
+        assertThat(formatSql(new Identifier("1", true))).isEqualTo("\"1\"");
+    }
+
+    @Test
+    public void testQuotedIdentifiers()
```
Comment (martint) [inline core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java:53]: This is misnamed and confusing. The tests are testing for identifiers that contain quotes, not what's commonly understood as a "quoted identifier". I would just fold these tests into the method above. No need for a separate one.

Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: Done

File: core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java (hunk 15 above)
Comment (nevillelyh) [reply core/trino-parser/src/test/java/io/trino/sql/TestSqlFormatter.java]: Done.
